INTRODUCTION

This dataset contains 5,001 records with facial feature measurements and gender labels. It includes attributes such as hair length, forehead dimensions, nose shape, and lip characteristics. The dataset can be used for gender classification based on facial features.

IMPORT LIBARIES AND LOAD DATASET 

In [2]:
#import libaries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import warnings
warnings.filterwarnings('ignore')
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn import metrics


#import dataset
df=pd.read_csv(r'C:\Data Science\data_set\gender.csv')
print(df)




      long_hair  forehead_width_cm  forehead_height_cm  nose_wide  nose_long  \
0             1               11.8                 6.1          1          0   
1             0               14.0                 5.4          0          0   
2             0               11.8                 6.3          1          1   
3             0               14.4                 6.1          0          1   
4             1               13.5                 5.9          0          0   
...         ...                ...                 ...        ...        ...   
4996          1               13.6                 5.1          0          0   
4997          1               11.9                 5.4          0          0   
4998          1               12.9                 5.7          0          0   
4999          1               13.2                 6.2          0          0   
5000          1               15.4                 5.4          1          1   

      lips_thin  distance_nose_to_lip_l

DATA CLEANING

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   long_hair                  5001 non-null   int64  
 1   forehead_width_cm          5001 non-null   float64
 2   forehead_height_cm         5001 non-null   float64
 3   nose_wide                  5001 non-null   int64  
 4   nose_long                  5001 non-null   int64  
 5   lips_thin                  5001 non-null   int64  
 6   distance_nose_to_lip_long  5001 non-null   int64  
 7   gender                     5001 non-null   object 
dtypes: float64(2), int64(5), object(1)
memory usage: 312.7+ KB


In [4]:
df.isna().sum()

long_hair                    0
forehead_width_cm            0
forehead_height_cm           0
nose_wide                    0
nose_long                    0
lips_thin                    0
distance_nose_to_lip_long    0
gender                       0
dtype: int64

In [5]:
df.duplicated().sum()

1768

In [6]:
df.drop_duplicates(inplace=True)

In [7]:
df.drop(columns=["distance_nose_to_lip_long"],inplace=True)

In [8]:
le=LabelEncoder()
df["gender"]=le.fit_transform(df["gender"])

EXTRACTING INDEPENDENT AND DEPENDENT VARIABLES

In [9]:
x=df[["long_hair","forehead_width_cm","forehead_height_cm","nose_wide","nose_long","lips_thin"]].values
x=pd.DataFrame(x)
y=df["gender"].values
y=pd.DataFrame(y)

SPLITING DATA INTO TRAIN AND TEST DATA

In [10]:
#split the dataset into training and testing sets
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.25,random_state=67)

FEATURE SCALING

In [11]:
#feature scalling 
st=StandardScaler()
x_test=st.fit_transform(x_test)
x_train=st.fit_transform(x_train)


MODEL BUILDING AND EVALUATE PREDICTIONS

MODEL OF KNN

In [12]:
from sklearn.neighbors import KNeighborsClassifier
classifier=KNeighborsClassifier(n_neighbors=5,metric='minkowski',p=2)
classifier.fit(x_train,y_train)

KNeighborsClassifier()

In [13]:
y_predict=classifier.predict(x_test)
print(y_test)
print(y_predict)

      0
2289  0
1789  1
925   1
1745  1
2330  1
...  ..
1610  1
2452  1
2155  1
82    0
695   0

[809 rows x 1 columns]
[0 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 0 0 1 0 1 0 1 1 1 1 1 0 1 0 1 0 1 0 0 1 1
 1 1 0 1 0 1 1 0 0 0 1 0 1 1 1 0 1 1 1 0 0 0 1 0 0 1 0 1 0 1 1 1 0 0 0 1 0
 0 0 0 1 1 1 0 1 1 0 1 1 1 1 0 0 0 1 0 1 1 1 0 0 1 1 0 1 0 1 1 0 1 1 1 0 1
 0 1 1 0 1 0 1 1 0 1 1 1 0 0 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 1 0
 0 1 1 1 1 1 1 1 0 1 0 0 1 0 1 0 0 0 1 0 1 1 1 0 1 1 0 1 1 0 1 0 1 0 1 1 1
 1 0 1 1 0 0 0 0 0 1 0 1 0 1 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 0
 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 0 0 1 1 0 1 0 1 0 1 1 1 0 0 1 0 1 0 0 0 1
 1 1 0 1 0 0 0 0 0 1 0 0 0 1 0 0 0 1 1 1 0 0 0 1 0 0 1 1 0 0 1 0 0 0 0 1 1
 0 0 0 0 1 1 1 1 0 0 0 1 1 1 0 0 1 1 0 1 0 1 0 0 1 1 1 1 0 1 1 0 1 1 0 0 1
 0 0 0 0 1 1 1 1 0 1 0 0 0 0 1 1 1 1 1 1 1 0 1 1 0 0 1 1 0 1 1 1 1 1 0 1 0
 1 1 0 0 1 1 0 1 1 1 0 1 1 0 1 0 1 1 1 0 1 0 1 0 1 1 1 1 1 1 1 1 1 1 0 1 0
 1 0 1 1 0 1 0 0 1 1 1 1 0 0 1 0 0 0 0 1 0 1 0 0 1 0 0 

In [14]:
print("mse value of knn regression:",metrics.mean_squared_error(y_predict,y_test))
print("accuracy in knn regression:",metrics.accuracy_score(y_predict,y_test)*100)

mse value of knn regression: 0.07045735475896168
accuracy in knn regression: 92.95426452410383


MODEL OF LOGISTIC REGRESSION

In [15]:
from sklearn.linear_model import LogisticRegression
model4=LogisticRegression(multi_class='multinomial',random_state=80)
model4.fit(x_train,y_train)

LogisticRegression(multi_class='multinomial', random_state=80)

In [16]:
y_predict=model4.predict(x_test)
print(y_test)
print(y_predict)

      0
2289  0
1789  1
925   1
1745  1
2330  1
...  ..
1610  1
2452  1
2155  1
82    0
695   0

[809 rows x 1 columns]
[0 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 0 0 1 0 1 0 1 1 1 1 1 0 1 0 1 0 1 0 0 1 1
 1 1 0 1 0 0 1 0 0 0 1 0 0 1 1 0 1 1 1 0 0 0 1 0 0 1 0 1 0 1 1 1 0 0 0 1 0
 0 1 0 1 1 1 1 1 1 0 1 1 1 1 1 0 0 1 0 1 1 1 0 0 1 1 0 1 0 1 1 0 1 1 1 1 1
 0 1 1 0 1 0 1 1 0 1 1 1 0 0 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 1 0
 0 1 1 1 1 1 1 0 0 1 0 0 1 0 1 0 0 0 1 0 1 1 1 0 1 1 0 1 1 0 1 0 1 0 1 1 1
 1 0 1 1 0 0 0 0 0 1 0 1 0 1 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 0
 1 1 1 0 0 1 0 0 1 1 1 1 0 1 1 0 0 0 1 1 0 1 0 1 0 1 1 1 0 0 1 0 1 0 0 0 1
 1 1 0 1 0 0 0 0 0 1 0 0 0 1 0 1 0 1 1 1 0 0 0 1 0 0 1 1 0 0 0 0 0 0 0 1 1
 0 0 0 0 1 1 1 1 0 0 0 1 1 1 0 0 1 1 0 1 0 1 0 0 1 1 1 1 0 1 1 0 1 1 0 0 1
 0 0 0 0 1 1 1 0 0 1 0 0 0 0 0 1 1 1 1 1 1 0 1 1 0 0 1 1 0 1 1 1 1 1 0 1 0
 1 1 0 0 1 1 0 1 1 1 0 1 1 0 1 0 1 1 1 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1 0
 1 0 1 1 0 1 0 0 1 1 1 1 0 0 1 1 0 0 0 1 0 1 0 0 1 0 0 

In [17]:
print("mse value of logistic regression:",metrics.mean_squared_error(y_predict,y_test))
print("accuracy in logistic regression:",metrics.accuracy_score(y_predict,y_test)*100)

mse value of logistic regression: 0.06056860321384425
accuracy in logistic regression: 93.94313967861557


MODEL OF SVM

In [18]:
from sklearn.svm import SVC
model1=SVC(kernel='linear',random_state=67)
model1.fit(x_train,y_train)

SVC(kernel='linear', random_state=67)

In [19]:
y_predict=model1.predict(x_test)
print(y_test)
print(y_predict)

      0
2289  0
1789  1
925   1
1745  1
2330  1
...  ..
1610  1
2452  1
2155  1
82    0
695   0

[809 rows x 1 columns]
[0 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 0 0 1 0 1 0 1 1 1 1 1 0 1 0 1 0 1 0 0 1 1
 1 1 0 1 0 1 1 0 0 0 1 0 0 1 1 0 1 1 1 0 0 0 1 0 0 1 0 1 0 1 1 1 0 0 0 1 0
 0 1 0 1 1 1 1 0 1 0 1 1 1 1 1 0 0 1 0 1 1 1 0 0 1 1 0 1 0 1 1 0 1 1 1 1 1
 0 1 1 0 1 0 1 1 0 1 1 1 0 0 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 1 0
 0 1 1 1 1 1 1 1 0 1 0 0 1 0 1 0 0 0 1 0 1 1 1 0 1 1 0 1 1 0 1 0 1 0 1 1 1
 1 0 1 1 0 0 0 0 0 1 0 1 0 1 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 0
 1 1 1 0 0 1 0 0 1 1 1 1 0 1 1 0 0 0 1 1 0 1 0 1 0 1 1 1 0 0 1 0 1 0 0 0 0
 1 1 0 1 0 0 0 0 0 1 0 0 0 1 0 1 0 1 1 1 0 0 0 1 0 0 1 1 0 0 1 0 0 0 0 1 1
 0 0 0 0 1 1 1 1 0 0 0 1 1 1 0 0 1 1 0 1 0 1 0 0 1 1 1 1 0 1 1 0 1 1 0 0 1
 0 0 0 0 1 1 1 1 0 1 0 0 0 0 1 1 1 1 1 1 1 0 1 1 0 0 1 1 0 1 1 1 1 1 0 1 0
 1 1 0 0 1 1 0 1 1 1 0 1 1 0 1 0 1 1 1 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1 0
 1 0 1 1 0 1 0 0 1 1 1 1 0 0 1 0 0 0 0 1 0 1 0 0 1 0 0 

In [20]:
print("mse value of Svm regression:",metrics.mean_squared_error(y_predict,y_test))
print("accuracy in Svm regression:",metrics.accuracy_score(y_predict,y_test)*100)

mse value of Svm regression: 0.0580964153275649
accuracy in Svm regression: 94.19035846724351


MODEL OF RANDOM FOREST

In [21]:
from sklearn.ensemble import RandomForestClassifier
model2=RandomForestClassifier(n_estimators=10,criterion='entropy',random_state=67)
model2.fit(x_train,y_train)


RandomForestClassifier(criterion='entropy', n_estimators=10, random_state=67)

In [22]:
y_predict=model2.predict(x_test)
print(y_test)
print(y_predict)

      0
2289  0
1789  1
925   1
1745  1
2330  1
...  ..
1610  1
2452  1
2155  1
82    0
695   0

[809 rows x 1 columns]
[0 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 0 0 1 0 1 0 1 1 1 1 1 0 0 0 1 0 0 0 0 1 1
 1 1 0 1 0 1 1 0 0 0 1 0 0 1 1 0 1 1 1 0 0 0 1 0 0 1 0 1 0 1 1 1 0 0 0 1 0
 0 0 0 1 1 1 0 0 1 0 1 1 1 1 1 1 0 1 0 1 1 1 0 0 1 1 0 1 0 1 1 0 1 1 1 1 1
 0 1 1 0 1 0 1 1 0 0 1 1 0 0 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 1 0
 0 1 1 1 1 1 1 1 0 1 0 0 1 0 1 0 0 0 1 0 1 1 1 0 1 1 0 0 1 0 1 0 1 0 1 1 1
 1 0 1 1 0 0 0 0 0 1 0 1 0 0 1 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 0
 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 0 0 1 1 0 1 0 1 0 1 1 1 0 0 1 0 1 0 0 0 1
 1 1 0 1 0 0 0 0 0 1 0 0 0 1 0 1 0 1 1 1 0 0 0 1 0 0 1 1 0 0 1 0 0 0 0 1 1
 0 0 0 0 1 1 1 1 0 0 0 1 1 1 0 0 1 1 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1 0 0 1
 0 0 0 0 1 1 1 0 0 1 0 0 0 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 1 1 1 1 1 1 0 1 0
 1 1 0 0 1 1 0 1 1 1 0 1 1 0 1 0 1 1 1 0 1 0 1 0 1 1 1 1 1 1 1 1 1 1 0 1 0
 1 0 1 1 0 1 0 0 1 1 1 1 0 0 1 1 0 0 0 1 0 1 0 0 1 0 0 

In [23]:
print("mse value of random forest:",metrics.mean_squared_error(y_predict,y_test))
print("accuracy in random forest:",metrics.accuracy_score(y_predict,y_test)*100)

mse value of random forest: 0.08405438813349815
accuracy in random forest: 91.59456118665018


MODEL OF DECISION TREE  

In [24]:
from sklearn.tree import DecisionTreeClassifier
model3=DecisionTreeClassifier(criterion='entropy',random_state=67)
model3.fit(x_train,y_train)




DecisionTreeClassifier(criterion='entropy', random_state=67)

In [25]:
y_predict=model3.predict(x_test)
print(y_test)
print(y_predict)

      0
2289  0
1789  1
925   1
1745  1
2330  1
...  ..
1610  1
2452  1
2155  1
82    0
695   0

[809 rows x 1 columns]
[0 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 0 0 1 0 1 0 1 1 1 1 1 0 0 0 0 0 0 0 0 1 1
 1 1 0 1 0 1 1 0 0 0 1 0 0 1 1 0 1 1 1 0 0 0 0 0 0 1 0 1 0 1 1 1 0 0 0 1 0
 0 0 0 1 1 1 0 0 1 0 1 1 1 1 0 1 0 1 0 1 1 1 0 0 1 1 0 1 0 1 1 0 1 1 1 1 1
 0 1 1 0 1 0 1 1 0 0 1 1 0 0 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 1 0
 0 1 1 1 1 1 1 1 0 1 0 0 1 0 1 0 0 0 1 0 1 1 1 0 1 1 0 0 1 0 1 0 1 0 1 1 1
 1 0 1 1 0 0 0 0 0 1 0 1 0 0 1 1 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0
 1 0 1 0 0 1 0 0 1 1 1 1 0 1 0 0 0 0 1 1 0 1 0 1 0 1 1 1 0 0 1 0 1 1 0 0 1
 1 1 0 1 0 0 0 0 0 1 0 0 0 1 0 0 0 1 1 1 0 1 0 1 0 0 1 1 0 0 1 0 0 0 0 1 1
 0 0 0 0 1 1 1 1 0 0 0 1 1 1 0 0 0 1 0 1 1 1 1 0 1 1 1 1 0 1 1 0 1 1 0 0 1
 0 0 0 0 1 1 1 0 0 1 0 0 0 0 1 1 1 1 0 0 1 0 1 1 0 0 1 0 1 1 1 1 1 1 0 1 0
 1 1 0 0 1 1 0 1 1 0 0 1 1 0 1 0 1 1 1 0 1 0 1 0 1 1 1 1 1 1 1 1 1 1 0 1 0
 1 0 1 1 0 1 0 0 1 1 1 1 0 0 1 1 0 0 0 1 1 1 0 0 1 0 0 

In [26]:
print("mse value of decision tree:",metrics.mean_squared_error(y_predict,y_test))
print("accuracy in decision tree:",metrics.accuracy_score(y_predict,y_test)*100)

mse value of decision tree: 0.09641532756489493
accuracy in decision tree: 90.3584672435105


SUMMARY

This project aimed to predict gender classification by using with given attributes such as hair length,forehead dimensions,nose shape,lip characteristics datas.After processing, we tested multiple model ,inculding Logistic regression ,Decision tree,Random forest,SVM,KNN.The SVM model achieved the highest accuracy of 94.19%,making it the best performing model. 